src/Wallpaper/Portal.vala

```diff
@@ -1,14 +1,114 @@
+       
+        var settings = new GLib.Settings(schema);
+        settings.set_string(key, value);
+        settings.apply();
+      
+        
+    }
+
+    public void set_background_wallpaper(string uri) throws Error {
+        set_gsettings(BACKGROUND_SCHEMA, "picture-uri", uri);
+    }
+    
+    public void set_lockscreen_wallpaper(string uri) throws Error {
+        set_gsettings(LOCKSCREEN_SCHEMA, "picture-uri", uri);
+    }
+    
+
     [DBus (name = "SetWallpaperURI")]
-	public async uint set_wallpaper_u_r_i(GLib.ObjectPath handle, string app_id, string parent_window, string uri, GLib.HashTable<string, GLib.Variant> options) throws DBusError, IOError {
-        debug ("set_wallpaper_u_r_i");
-        return 0;
+    public async uint set_wallpaper_uri(GLib.ObjectPath handle, string app_id, string parent_window, string uri, GLib.HashTable<string, GLib.Variant> options, out uint response) throws DBusError, IOError {
+        debug("set_wallpaper_uri");
+
+
+        string get_uri="";
+
+        try {
+            get_uri = Uri.parse(uri, UriFlags.NONE).get_path();
+        } catch (GLib.UriError e) {
+            stderr.printf("Error parsing URI: %s\n", e.message);
+        }
+
+        bool show_preview = false;
+        string set_on = "";
+        if (options != null) {
+            GLib.Variant? showPreviewVariant = options.lookup("show-preview");
+            if (showPreviewVariant != null && showPreviewVariant.get_boolean()) {
+                show_preview = true;
+            }
+
+            GLib.Variant? setOnVariant = options.lookup("set-on");
+            if (setOnVariant != null) {
+                set_on = setOnVariant.get_string();
+            }
+        }
+
+        debug("show-preview: %s", show_preview ? "true" : "false");
+        debug("set-on: %s", set_on);
+
+        if (show_preview) {
+            try{
+                show_preview_image(get_uri,parent_window);
+            } catch (GLib.Error e) {
+                stderr.printf("Error showing preview: %s\n", e.message);
+                return response=1;
+            }
+        }
+
+        uint response_code = 0;
+
+        if (set_on == "background") {
+            try {
+                set_background_wallpaper(get_uri);
+                response_code = 0;
+            } catch (GLib.Error e) {
+                response_code = 1;
+            }
+        } else if (set_on == "lockscreen") {
+            try {
+                set_lockscreen_wallpaper(get_uri);
+                response_code = 0;
+            } catch (GLib.Error e) {
+                response_code = 1;
+            }
+        } else if (set_on == "both") {
+            try {
+                set_background_wallpaper(get_uri);
+                set_lockscreen_wallpaper(get_uri);
+                response_code = 0;
+            } catch (GLib.Error e) {
+                response_code = 1;
+            }
+        } else {
+            response_code = 2;
```

Comment:
question: what do these integer literals mean?  can we extract to constants for better readability?

Reply:
I employed it in accordance with the xdg desktop documentation : https://flatpak.github.io/xdg-desktop-portal/docs/doc-org.freedesktop.portal.Request.html#org-freedesktop-portal-request

0: Success; the request is executed.
1: The user cancelled the interaction.
2: The user interaction concluded in an alternate manner.

If it's preferable to use constants to represent these outcomes, I'll make the necessary changes.

Reply:
Hmm, maybe just put a link to that URL as a comment above in case we need to know about it in the future.